### ISSUE TITLE
revise label (on staffGrp and staffDef)

### ISSUE THREAD

**Issue by rettinghaus** [music-encoding/music-encoding#361]
`@label` simply 
> Provides a name or label for an element.

Question is: when used on `<staffGrp>` and `<staffDef>` they carry the *printed name* of the staff, that one can find before the start of the line. But this way there is no possibility to express that the given label is a non-visible (e.g. in Verovio it has to be blank). 
Other way round: Why should a `<staff label="Tenor"/>` gets printed, and a `<note label="most beautiful moment in the whole piece"/>` not? (Or where would be the intended place to put it?)
This is tricky! 
We could introduce a `@label.vis`, but then we would need a `@label.abbr.vis`, too. (And that would blow up the attribute zoo on `staffDef` even more.) 
When `@type` becomes a common attribute, we could drop `@label` from the list, but as they (may) carry different meanings, that wouldn't be very wise. 
Any ideas?

**Comment by pe-ro** [comment]
@rettinghaus, I think we need to move away from `@label` as the printable/displayable text associated with an object, in this case `<staffGrp>` or `<staffDef>`.  In these cases, the child element `<label>` should be used instead.  `@label` should be used for the name/label/appelation associated with the XML element, kind of like an ID, but less constrained.  To reinforce this, `@label` should be renamed, but I'm not certain yet what to call it.  "tooltip" is on the right track, but maybe too specific.  In HTML `@alt` serves this purpose for images.

**Comment by rettinghaus** [comment]
@pe-ro I basically like `@label` like it is right now. IMHO it is more like `@title` than `@alt` in HTML and it goes wonderful alongside with `@type` as a somewhat more human-readable name string. 
Perhaps we could/should drop the label element and introduce a `<partName>` (or something) instead for `<staffGrp>` and `<staffDef>`?

**Comment by pe-ro** [comment]
It's the "or something" that gives me pause -- So far, I've been unsuccessful in coming up with an element name that is more specific than `<label>` but generic enough to be used for both `<staffGrp>` and `<staffDef>`.  `<partName>` isn't a good choice, IMO, as neither a staff nor a staff group always represent a so-called "part" -- something which MEI doesn't actually have any conception of.

Another solution (as opposed to renaming `@label`) would be to introduce 2 new elements -- `<staffName>` and `<staffGrpName>`.  This identifies each "name" very specifically, but introduces a different inconsistency, since other elements still use `<label>` for the same purpose.  I think this inconsistency is worse than the current situation since it involves many more elements and it means that the content models of `<staffDef>` and `<staffGrp>` have to be treated as special cases.

**Comment by lpugin** [comment]
Since I am sure you thought about it, why are you discarding `<name>` as an option?

**Comment by pe-ro** [comment]
You're right, I did think about it.  Because of its definition -- "Proper noun or noun phrase." -- it doesn't seem to fit this purpose.

**Comment by pe-ro** [comment]
`<label>` is more appropriate because, well, the string attached to the left-hand side of a staff or staffGrp is a label.

**Comment by pe-ro** [comment]
It's the `@label` that is think is named inappropriately.

**Comment by pe-ro** [comment]
I misspoke, `@label` isn't named inappropriately, but rather confusingly -- `<label>` contains (or ought to contain) text that's in the document, `@label` is for text/a name associated with the element itself, similar to `@title` in HTML. (Thanks to @rettinghaus for pointing that out.)

Just in case you're wondering, I don't think `@title` will work as a replacement for `@label`.  There are places where the name "title" will lead to confusion.

**Comment by kepper** [comment]
I'd like to add measure to the list of elements to consider here. Very frequently, measures have numbers which aren't numbers, like 21a / 21b etc. Is that a sufficiently "number-like designation for an element" to be stored in @n (where this description comes from), or does it require (current) @label? If so, how do the considerations on this issue affect that situation?

**Comment by pe-ro** [comment]
Yes, '21a', etc. is sufficiently number-like to be placed in `@n`.  A major consideration is that these things are commonly called "measure numbers" (whether they're actually numeric or not), not "measure labels". :-)

The only "correct" solution is to provide `<number>` as a child of `<measure>`, `<staffName>` and `<staffGrpName>` as children of `<staffDef>` and `<staffGrp>`, and a specialized element for every other place where such a thing is necessary.  But, I doubt that we could ever resolve the chain of new conflicts this would create.  In addition, we'd be complicating every one of those content models, effectively making each of them a special case.

**Comment by lpugin** [comment]
I cannot find where this was discussed, but I think there was the idea to have two types of attribute classes for `@n`. One for number-like such as above (e.g., '21a') and one for proper number to be used for elements such as `staff` or `layer`.

**Comment by pe-ro** [comment]
And that's what was done in the last revision -- att.common defines `@n` as number-like and `<layer>`, `<staff>`, and `<staffDef>` each define their own `@n` to be an integer. There's no class definition for `@n` as an integer.  But, there's an att.numberName definition of `@n` used by those elements that don't participate in att.common.

**Comment by lpugin** [comment]
Great. 

> There's no class definition for @n as an integer.

Can't we have one ;-) ? Attribute without class definition do not make it through the LibMEI version used in Verovio. Since it seems that we have a very few of them it would make it more consistent to have attribute classes everywhere - unless there is a good reason not to have it, of course, in which case we can think how to change LibMEI.

**Comment by pe-ro** [comment]
I was just looking at that.  I think creating a class (att.nInteger) is a good thing.  Will also rename att.numberName to att.nName and use it as part of att.common.

**Comment by pe-ro** [comment]
While I'm revising these classes, do I hear a motion to change the number-like definition of `@n` to force it to contain digits at the beginning?  That is, "21" or "21a" would be acceptable, but "a21" would not and would have to be recorded in `@label`.  I've been reluctant to do this before because it doesn't really resolve what a label is.  This is why I suggested renaming `@label`.

**Comment by rettinghaus** [comment]
@pe-ro A little bit off the track, but I like the idea of having a `<number>` inside of `<measure>` as this could be a way to solve the issue here: rism-ch/verovio#177

**Comment by pe-ro** [comment]
@rettinghaus, I understand.  And maybe it's time we make sure attributes are never used for document content.  But, I've been reluctant to start that TEI-like "war on attributes" as it will consume many things in its fury.  I don't know if it can be waged piecemeal.

**Comment by pe-ro** [comment]
@lpugin, attributes that aren't defined by a class aren't handled by LibMEI at all, or just not by the version you're using in Verovio?

**Comment by lpugin** [comment]
Only in the version used by Verovio. The original version also generate C++ classes for the elements. In Verovio, those are defined by hand.

**Comment by pe-ro** [comment]
Hmmm.  Sounds like a Verovio issue.  ;-)  Why aren't you using the original version?

**Comment by pe-ro** [comment]
@rettinghaus, Looking more carefully, I see that what's really under discussion in rism-ch/verovio#177 is the more-or-less global rendering of bar numbers, not bar numbers on individual measures.  That kind of thing really should be handled higher up in the hierarchy, first by `@mnum.visible` as you suggested.  I think adding another attribute, something like `@mnum.pattern` with values like those suggested by Niko, would be a good 2nd step.

**Comment by lpugin** [comment]
@pe-ro : the original version of LibMEI produces element classes that do nothing else than reading and writing MEI. So using this would have meant for each element class writing by hand an additional child class to add rendering stuff. This did not look like a very good design...

**Comment by pe-ro** [comment]
Makes sense.

**Comment by pe-ro** [comment]
Another question -- How much trouble does it cause when attribute classes change names?  I imagine that it's not insignificant, but sometimes there's just no other way to handle a problem.

**Comment by lpugin** [comment]
Changing attribute classes name is not much work.

**Comment by pe-ro** [comment]
I'm glad to hear that, since changing class names is useful for clarification of attribute function.

**Comment by bwbohl** [comment]
General remark: We should not confuse issues derived from development of a certain software dealing with MEI files and those issues emerging from our data format itself!

Please don't misunderstand, I do not want to narrow the highly useful hints for MEI that come from the increased use and development of Verovio!!

Some statements to the issues raised in the above discussion:

1. creating an attribute class att.nInteger
I think it most useful to always define classes for attributes (and elements, of course). This makes for a consistent way of defining the schema.

2. Elements vs attributes OR the attribute war
Generalising that only elements should hold data that's in "the document" being encoded is not a way we can go with MEI.
* First this would issue a huge break in backward compatibility as currently many attributes are used to capture data, e.g. `@pname`, `@oct`, `@dur` on `<note/>`. If you do not agree on this, I still would   shy away from a discussion about this as it will result in a huge philological debate trying to define what is "data" in this sense and what not.
* What if we're not encoding a pre-existing document but the MEI itself is the born-digital document? All that would be in this "document" then would have to beb encoded in elements ;-)

3. multiple definitions of the same attribute
concerns "1."
Don't you consider it "bad style" modelling two attributes with different definitions but the same name in the same namespace? The qualified identifier would be the same and not resolvable `http://www.music-encoding.org/ns/mei:@n`

4. current `@n`s
What is a number-like designation for an element?
[0-9]+[^"]*
that correct? Or are there more restrictions in rng:tokens?

What is a non-negative integer value functioning as a "name"?
In other words a positive number 1 (or 0?) to unlimited

Can you help me to understand the purpose of this distinction?

Just a couple of cents and some follow up questions ;-)

**Comment by pe-ro** [comment]
>General remark: We should not confuse issues derived from development of a certain software dealing with MEI files and those issues emerging from our data format itself!
>
>Please don't misunderstand, I do not want to narrow the highly useful hints for MEI that come from the increased use and development of Verovio!!

We're in total agreement.  But if something is difficult to implement, then it may be the case that underlying assumptions may be incorrect.  Iterative development of the schema and an implementation stands a good chance of working out the kinks in both. :-)

> creating an attribute class att.nInteger
>I think it most useful to always define classes for attributes (and elements, of course). This makes for a consistent way of defining the schema.

I agree, but "always" is a high standard that can't always be met.

> Elements vs attributes OR the attribute war
> Generalising that only elements should hold data that's in "the document" being encoded is not a way we can go with MEI.

You're correct re: backward compatibility, but backward compatibility must be secondary to the improvement (more targeted, easier to understand, all-around "better") of the conceptual model.

> I still would shy away from a discussion about this as it will result in a huge philological debate trying to define what is "data" in this sense and what not.

The distinction between content and meta-content in music notation is not as clear as in text documents.  The usual text-based approach of "the text is what remains after the meta-content; i.e., markup, is removed" isn't applicable.  I think we agree on that.  But, what I was getting at with the phrase "war on attributes", clumsily I must add, was that we have to give a great deal of thought about the "levels of existence" in music notation.  By this I mean, is the thing we're talking about a so-called "first-class" object; that is, it may have attributes, or is it a "second-class" object; that is, it's simply an attribute of something else.  A good example here is the number  "3" displayed along with a tuplet.  One can think of this number as a property of the tuplet, in which case it can be represented by an attribute -- `@num="3"`.  On the other hand, if the number has properties of its own -- size, color, position, for example -- then it should be represented by an element.  The problem is that the needs of encoders/consumers of an encoding of music notation vary so widely that there is no single, correct approach.  I was simply suggesting that we keep the tension between the two possibilities in mind and that we could move in the direction of using elements more often.  I'm aware that this might cause compatibility issues, but who cares?  I don't.  :-)

> Don't you consider it "bad style" modelling two attributes with different definitions but the same name in the same namespace? The qualified identifier would be the same and not resolvable http://www.music-encoding.org/ns/mei:@n

It shouldn't be over-used, but I don't consider it "bad style".  I think it's a compromise that has to be made sometimes to accommodate the contextual nature of document encoding.  For me, the various forms of `@n` are components of a larger definition -- "name or number" -- where, in fact, number is often functioning like a name, like when a ship captain refers to his immediate subordinate as "No. 1".  The creation of multiple specifications for `@n` are a way to enforce the various "situational" values that it can take -- sometimes it's expected to be an integer, sometimes an integer with optional alphabetic "extensions" (like when measures are "numbered" 1, 2a, 2b, etc.).

`http://www.music-encoding.org/ns/mei:@n` is contradictory to this approach.  But my approach is in line with human understanding; that is, we sometimes give dissimilar things the same name.  It's better to make the schema understandable by humans than machines.

> What is a number-like designation for an element?
> [0-9]+[^"]*
> that correct? Or are there more restrictions in rng:tokens?
>
> What is a non-negative integer value functioning as a "name"?
> In other words a positive number 1 (or 0?) to unlimited

The next revision of the schema provides 3 definitions of `@n`.  The most generic of these is --
```
<classSpec ident="att.tokenName" module="MEI.shared" type="atts">
  <desc>Attributes used to supply a token-like name for an element.</desc>
  <attList>
    <attDef ident="n" usage="opt">
      <desc>Provides a token-like designation for an element; i.e., one that contains a single
        token.</desc>
      <datatype>
        <rng:data type="token">
          <rng:param name="pattern">(\p{L}|\p{N}|\p{P}|\p{S})*</rng:param>
        </rng:data>
      </datatype>
    </attDef>
  </attList>
</classSpec>
```
in other words, a token consisting of letters, numbers, punctuation and space characters.  Since spaces are allowed, the entire value of `@n` is a unitary thing, for example, "my Dog".  This is a way of saying that despite the usual XML practice of using spaces to separate an attribute value string into multiple values, there's only one value here.

The second "version" of `@n` is more restrictive --
```
<classSpec ident="att.numberName" module="MEI.shared" type="atts">
  <desc>Attributes used to supply a number-like name for an element.</desc>
  <attList>
    <attDef ident="n" usage="opt">
      <desc>Provides a number-like designation for an element; i.e., one that contains one or more
        digits.</desc>
      <datatype>
        <rng:data type="token">
          <rng:param name="pattern">.*(\d)+.*</rng:param>
        </rng:data>
      </datatype>
    </attDef>
  </attList>
</classSpec>
```
This enforces a restriction that the number/name must contain a digit.  This would allow "No. 1" or "21a", but not "sixty-two".  This is particularly useful for measure "numbers", which usually have one or more digits, but aren't necessarily composed of digits exclusively.

The final version is most restrictive --
```
<classSpec ident="att.integerName" module="MEI.shared" type="atts">
  <desc>Attributes used to supply an integer number designation for an element.</desc>
  <attList>
    <attDef ident="n" usage="opt">
      <desc>A non-negative integer value functioning as a "name".</desc>
      <datatype>
        <rng:data type="nonNegativeInteger"/>
      </datatype>
    </attDef>
  </attList>
</classSpec>
```
This limits the value to integers from 0 to infinity.  A "number functioning as a name" is designed to accommodate "names" like "7" and "99", which were the names of characters in "Star Trek" and "Get Smart", respectively, but also the "names" of staves. :-)

**Comment by bwbohl** [comment]
> On 11 Jan 2017, at 17:58, pe-ro <notifications@github.com> wrote:
> 
>> General remark: We should not confuse issues derived from development of a certain software dealing with MEI files and those issues emerging from our data format itself!
>> 
>> Please don't misunderstand, I do not want to narrow the highly useful hints for MEI that come from the increased use and development of Verovio!!
>> 
> We're in total agreement. But if something is difficult to implement, then it may be the case that underlying assumptions may be incorrect. Iterative development of the schema and an implementation stands a good chance of working out the kinks in both. :-)

Perfect argument, nevertheless we we should not model the way a software implements sth. but in a way that most precisely describes the musical fact. Maybe we should try to define more closely which domain MEI files really encode, cf. #318

>> 
>> creating an attribute class att.nInteger
>> I think it most useful to always define classes for attributes (and elements, of course). This makes for a consistent way of defining the schema.
>> 
> I agree, but "always" is a high standard that can't always be met.

Well, yes and no, of course the extreme result could be a class for each attribute ;-) And we do not want that. Nevertheless ODD should allow to `@exclude` class members in the reference e.g. of a a attribute class referenced from an element definition. Maybe implementing a “don't yet know where this should belong”-Class?

> 
>> Elements vs attributes OR the attribute war
>> Generalising that only elements should hold data that's in "the document" being encoded is not a way we can go with MEI.
> 

Concerning your two statements: yes we do agree and better is better even if it breaks backward compatibility!


>> Don't you consider it "bad style" modelling two attributes with different definitions but the same name in the same namespace? The qualified identifier would be the same and not resolvable http://www.music-encoding.org/ns/mei:@n
>> 
> It shouldn't be over-used, but I don't consider it "bad style". I think it's a compromise that has to be made sometimes to accommodate the contextual nature of document encoding. For me, the various forms of @n are components of a larger definition -- "name or number" -- where, in fact, number is often functioning like a name, like when a ship captain refers to his immediate subordinate as "No. 1". The creation of multiple specifications for @n are a way to enforce the various "situational" values that it can take -- sometimes it's expected to be an integer, sometimes an integer with optional alphabetic "extensions" (like when measures are "numbered" 1, 2a, 2b, etc.).
> 
> http://www.music-encoding.org/ns/mei:@n is contradictory to this approach. But my approach is in line with human understanding; that is, we sometimes give dissimilar things the same name. It's better to make the schema understandable by humans than machines.

I don’t see the contradiction. Yes human understandability will be necessary as long as humans develop and use the schema ;-) Nevertheless, having a single global definition of `@n` is something that humans AND machines might prefer.
Bringing together these two desiderata might be reached by adding to this single definition additional schematron rules (only throwing warnings not errors) covering the three usage cases. One could even consider the different cases user specific, so this approach would provide defined spots for customisation.

**Comment by lpugin** [comment]
> Perfect argument, nevertheless we we should not model the way a software implements sth. but in a way that most precisely describes the musical fact.

@bwbohl : I agree too, but I am not sure what in this discussion is in your opinion biased towards a certain tool.

**Comment by bwbohl** [comment]
> Perfect argument, nevertheless we we should not model the way a software implements sth. but in a way that most precisely describes the musical fact.
> 
> @bwbohl : I agree too, but I am not sure what in this discussion is in your opinion biased towards a certain tool.

It was more as a general remark or caveat, as the question arose from the problem of how to encode labels but make them invisible in a verovio rendering.
The discussion has turned out perfectly fine ;-)

**Comment by pe-ro** [comment]
> I don’t see the contradiction. Yes human understandability will be necessary as long as humans develop and use the schema ;-) Nevertheless, having a single global definition of `@n` is something that humans AND machines might prefer. 
> Bringing together these two desiderata might be reached by adding to this single definition additional schematron rules (only throwing warnings not errors) covering the three usage cases. One could even consider the different cases user specific, so this approach would provide defined spots for customisation.

Either way -- multiple definitions of `@n` or a single definition of `@n` with context-dependent schematron rules -- the definition of `@n` ends up being contextual.

**Comment by pe-ro** [comment]
I've reopened #378 because I'm not satisfied with my earlier "solution".  @bwbohl's comments got me thinking about consistency and the lack thereof in my attempts to resolve the conflict between `@n` and `@label`.  I now believe the major sticking point is trying to retain TEI's definition of `@n` as a "number or a name".  For MEI it makes more sense to consistently define `@n` as a number and allow `@label` to take the role of "name".

My goal is to make `@n` a nonNegative integer in almost every case.  I say "almost every case" because some "numbers", like those used to designate measures and page breaks, aren't always integers.  However, because they contain digits and other characters but not spaces, they are sufficiently "number-like" to fit the general definition of `@n` as a "number".  The integer vs. number-like nature of `@n` will be enforced via two attribute classes --
```
<classSpec ident="att.nInteger" module="MEI.shared" type="atts" mode="add">
  <desc>Attributes used to supply an integer number designation for an element.</desc>
  <attList>
    <attDef ident="n" usage="opt">
      <desc>A non-negative integer.</desc>
      <datatype>
        <rng:data type="nonNegativeInteger"/>
      </datatype>
    </attDef>
  </attList>
</classSpec>

<classSpec ident="att.nNumberLike" module="MEI.shared" type="atts" mode="add">
  <desc>Attributes used to supply a number-like designation for an element.</desc>
  <attList>
    <attDef ident="n" usage="opt">
      <desc>Provides a number-like designation for an element; i.e., one that contains at
        least one digit and other characters but no spaces.</desc>
      <datatype>
        <rng:data type="token">
          <rng:param name="pattern"
            >(\p{L}|\p{P}|\p{S})*(\p{N})+(\p{L}|\p{P}|\p{S}|\p{N})*</rng:param>
        </rng:data>
      </datatype>
    </attDef>
  </attList>
</classSpec>
```
With `@n` consistently containing a number or number-like value, the function of `@label` becomes clearer; that is, to contain a name/label (i.e., a non-numeric string perhaps including spaces) associated with the element.  I think there are 2 likely uses of this attribute -- to capture a "tooltip", which, as @rettinghaus has pointed out, is similar to HTML's "title" attribute, or to record a display label for those elements that can't contain `<label>`.  This last group mainly consists of metadata elements which often need a preceding label to make sense, for example
```
Song Title: Wo ist Sylvia's Mother?
Composer: Frank Schubert's Medicine Show
```
Of course, these labels can often be generated based on other features in the markup, but sometimes, esp. when one is re-using data from another source, an existing catalog for example, it's handy to have a place to store a label that source data may have carried.

**Comment by pe-ro** [comment]
Having (I hope) disambiguated `@n` and `@label`, we can return to Klaus's original issue -- the use of `@label` and `@label.abbr` for content; that is, the string that appears as the label/name of a `<staff>` or `<staffGrp>`.

I believe it would be best to discourage the use of attributes for this purpose.  Now that we've nailed down the definition of `@label` as a name/identifying string for an **element**, it cannot be used to capture **content**.

So, I think `@label.abbr` should be removed and users directed to use the child `<label>` element of the elements where it was formerly present; that is, `<layerDef>`, `<staffDef>` and `<staffGrp>`.  To differentiate between those labels which appear on the first page of a score and those that occur later, either another element can be created, something like `<labelAbbr>` or `<label @type>` can be employed.  I favor creating a new element, since `@type` would affect the use of `<label>` in other locations, such as in `<list>` where a value of `@type='abbr'` wouldn't make sense.

I hesitate to suggest that `@label` also be removed on these elements (layerDef, staffDef, and staffGrp) to force the use of `<label>`, but it's something that could be considered later or in another context; that is, MEI Go.

**Comment by pe-ro** [comment]
Doh!  Actually, proceeding from the generic to the specific, `@n` should be defined as a TOKEN by default, and as a NonNegativeInteger in special cases.

**Comment by lpugin** [comment]
Abbreviated versions of the label are currently `<label2>`. Maybe we should change it to something more explicit such as `<labelAbbr>` or `<label abbr="true">`?

**Comment by pe-ro** [comment]
`<label2>` was chosen because of its similarity to `<pgHead2>` and `<pgFoot2>` but `<labelAbbr>` is fine too.

`<labelAbbr>` is better than `<label abbr="true">` since the latter 1) introduces the possibility of a value other than "true" and 2) may lead to confusion between `@abbr` and `@type` and `@class`.

**Comment by lpugin** [comment]
`<label2>` looks like a programmer variable choice and is not very explicit. Also I find that having an element just for the label abbreviation is over the top, hence the proposal with an attribute. But I am fine with `<labelAbbr>` so let us go for that.

**Comment by craigsapp** [comment]
Yes, `<labelAbbr>` is more obvious as to what it is than `<label2>`.

**Comment by pe-ro** [comment]
Glad you approve :-)

**Comment by craigsapp** [comment]
It happens once in a while :-)

**Comment by rettinghaus** [comment]
`<labelAbbr>` is fine, but I was wondering if not `<labelRpt>` would fit better into naming scheme. In the end it is a container for _repeating the label_. Could be less confusing if anybody asks why `<label>` goes over `<labelAbbr>` on the first line …

**Comment by kepper** [comment]
However, *Rpt is used differently in MEI so far: for mRpt etc. it denotes *an instance* of the preceding element (the label in our current case). What we really want here is something that explains *how repetitions look like in general*, not every single repetition. That's still not exactly what regular abbreviations do in MEI, but still, labelAbbr seems to be more appropriate than labelRpt to me…

**Comment by pe-ro** [comment]
@kepper, I concur.

**Comment by pe-ro** [comment]
I believe the major issues here have been resolved.

**Comment by rettinghaus** [comment]
What still is missing is a way to carry the (visual information of) labels, but don't show them in rendering. I wonder if `@visible` could do the trick on `<label>` and `<labelAbbr>`. 
(Although in *Verovio* this doesn't mean "don't print it" but "print it and make it invisible by default").

**Comment by pe-ro** [comment]
I can make `<label>` and `<labelAbbr>` members of att.visibility.  But first I'd like to know, in what circumstance is "print it and make it invisible by default" necessary?
